rust/lance/src/dataset/fragment.rs

```diff
@@ -1394,29 +1350,82 @@ impl FragmentReader {
                     // Apply ? inside the task to keep read_tasks a simple iter of futures
                     // for try_join_all
                     let projection = projection?;
-                    // We always get the row_id from the first reader and so we need that even
-                    // if the projection is empty
-                    let need_for_row_id = self.with_row_id && reader_idx == 0;
-                    if projection.fields.is_empty() && !need_for_row_id {
+                    if projection.fields.is_empty() {
                         // The projection caused one of the data files to become
                         // irrelevant and so we can skip it
                         Result::Ok(None)
                     } else {
                         Ok(Some(
                             reader
-                                .read_batch(
-                                    batch_id as i32,
-                                    params,
-                                    &projection,
-                                    self.deletion_vec.as_ref().map(|dv| dv.as_ref()),
-                                )
+                                .read_batch(batch_id as i32, params, &projection)
                                 .await?,
                         ))
                     }
                 }
             });
-        let batches = try_join_all(read_tasks).await?;
-        let batches = batches.into_iter().flatten().collect::<Vec<_>>();
+            let results = try_join_all(read_tasks).await?;
+            results.into_iter().flatten().collect::<Vec<RecordBatch>>()
+        } else {
+            let expected_rows = params
+                .clone()
+                .into()
+                .slice(0, rows_in_batch)
+                .unwrap()
+                .to_offsets()?
+                .len();
+            let dummy_array = Arc::new(NullArray::new(expected_rows));
+            vec![RecordBatch::try_new(
+                Arc::new(ArrowSchema::new(vec![ArrowField::new(
+                    "dummy",
+                    DataType::Null,
+                    false,
+                )])),
+                vec![dummy_array],
+            )
+            .unwrap()
+            .project(&[])
+            .unwrap()]
```

Comment:
I'm confused.  What is this doing?  Maybe add a comment?

Reply:
This is creating an empty record batch with the correct length. This is the simplest way I could think of, since `RecordBatch::empty()` is for zero rows and `RecordBatch::try_new()` errors if you pass no columns.